### ISSUE TITLE
Memory Leak During Prediction with `{training: true}`

### ISSUE THREAD

**Issue by AmitMY** [tensorflow/tfjs#7983]
**System information**
- TensorFlow.js version: 4.10.0

**Describe the current behavior**
There has been lots of discussion about model prediction with correct batchnorm norms (https://github.com/tensorflow/tfjs/issues/3152) which comes down to:
```ts
model.apply(tensor, {training: true})
```
Instead of 
```ts
model.apply(tensor)
```

In my specific model, this is leaking 2 tensors

**Describe the expected behavior**

No memory leaks regardless of training situation.

**Standalone code to reproduce the issue**

Fully standalone: https://drive.google.com/file/d/1DB9UzPDpZ8umTwsA8dvlWElCkU2ykLyw/view?usp=sharing
(to reproduce, I ran http-server .)

```ts
const model = await tf.loadLayersModel('model.json');

const tensor = tf.zeros([1, 1, 256, 256, 3]).toFloat();
tensor.print();

// Must apply model in training=True mode to avoid using aggregated norm statistics
const beforeTensors = tf.memory().numTensors
const pred = tf.tidy(() => model.apply(tensor, {training: true}));
console.log('leaking', tf.memory().numTensors - beforeTensors - 1, 'tensors');
// leaking 2 tensors

const beforeTensors2 = tf.memory().numTensors
const pred2 = tf.tidy(() => model.apply(tensor));
console.log('leaking', tf.memory().numTensors - beforeTensors2 - 1, 'tensors');
// leaking 0 tensors
```

**Comment by gaikwadrahul8** [comment]
Hi, @AmitMY 

Thank you for bringing this issue to our attention and I tried to replicate the same issue from my end and I'm also getting the same result which you've mentioned above even I tried with previous versions `@tensorflow/tfjs@4.10.0`, `@tensorflow/tfjs@4.9.0` but issue still exists so we'll have to dig more into this issue and will update you soon. Thank you!

I have added screenshot below for reference :

![image](https://github.com/tensorflow/tfjs/assets/115997457/809040cf-9a2c-495b-93cf-0335a8d3bdcd)

**Comment by borodadada** [comment]
There has always been a memory leak, I have a training cycle, the same thing over and over again, each time the size of the program in memory increases, and so does the execution time. Take the simplest model training, any, put it in a loop for 12 hours, and everything will be visible.

**Comment by gaikwadrahul8** [comment]
Hi, @AmitMY, @borodadada 

I apologize for the delayed response and as per my current understanding when the `model.apply()` function is called with `{training: true}`, it internally creates new tensors that are not disposed of within the `tidy` scope.

This leads to a memory leak because the intermediate tensors created by `model.apply()` with `{training: true} `are not cleaned up properly. As a result, the number of tensors in memory increases, leading to memory exhaustion over time if this pattern is repeated and to avoid the memory leak, you can modify the code to explicitly dispose of the tensors created by `model.apply()` with `{training: true}` using the `tf.dispose()` function.

If I have missed something here please let me know. Thank you.

```
async function main() {
        const model = await tf.loadLayersModel('web_model/model.json');
        const tensor = tf.zeros([1, 1, 256, 256, 3]).toFloat();
        tensor.print();

        const beforeTensors = tf.memory().numTensors;
        const pred = tf.tidy(() => {
            const prediction = model.apply(tensor, { training: true });
            tf.dispose([tensor, prediction]); // Dispose of intermediate tensors
            return prediction;
        });
        console.log('leaking', tf.memory().numTensors - beforeTensors - 1, 'tensors');
        // leaking 0 tensors
         }
         
    main();
```

**Comment by borodadada** [comment]
why can’t you add an initialization function, run it every time after training is completed, without input data, just a function, or I don’t understand something

**Comment by Varun-S10** [comment]
Hi @AmitMY, Thank you for bringing this to our attention and for providing a comprehensive overview. We appreciate the time you invested in creating this issue.

Currently, our internal development team is concentrated on other scheduled milestones, so we cannot prioritize this task immediately. However, we would love to see this move forward as a community-led effort.

If you’d like to submit a fix, please fork the repo and open a PR following the [TensorFlow Contribution Guidelines](https://www.tensorflow.org/community/contribute), [Tfjs Contributions ](https://github.com/tensorflow/tfjs/blob/master/tfjs-vis/CONTRIBUTING.md#how-to-contribute)and be sure to link this issue in your PR to avoid duplicate work, kindly leave a comment here if you plan to take this on.

Thank you again for your interest and time.

**Comment by Varun-S10** [comment]
Hi @AmitMY, Thank you for opening this issue. Since this issue has been open for a long time, we are closing this issue for now.
If this is still relevant and you would like to work on it, please feel free to reopen the issue or submit a pull request referencing it. Thank you for your understanding.
